src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1052,32 +1068,30 @@ ucs_status_t uct_ud_ep_flush_nolock(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
     }
 
     /* Expedite acknowledgment on the last skb in the window */
-    if (uct_ud_ep_is_last_ack_received(ep)) {
+    if (uct_ud_ep_is_last_ack_received(ep) &&
+        ucs_likely(!(flags & UCT_FLUSH_FLAG_CANCEL))) {
         uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_ACK_REQ);
-    } else {
-        uct_ud_ep_assert_tx_window_nonempty(ep);
-        skb = ucs_queue_tail_elem_non_empty(&ep->tx.window, uct_ud_send_skb_t,
-                                            queue);
-        if (!(skb->flags & UCT_UD_SEND_SKB_FLAG_ACK_REQ)) {
-            /* If we didn't ask for ACK on last skb, send an ACK_REQ message.
-             * It will speed up the flush because we will not have to wait until
-             * retransmit is triggered.
-             * Also, prevent from sending more control messages like this after
-             * first time by turning on the flag on the last skb.
-             */
-
-            /* Since the function can be called from the arbiter context it is
-             * impossible to schedule a control operation. So just raise a
-             * flag and if there is no other control send ACK_REQ directly.
-             *
-             * If there is other control arbiter will take care of it.
-             */
-            ep->tx.pending.ops |= UCT_UD_EP_OP_ACK_REQ;
-            if (uct_ud_ep_ctl_op_check_ex(ep, UCT_UD_EP_OP_ACK_REQ)) {
-                uct_ud_ep_do_pending_ctl(ep, iface);
-            }
+    } else if (!uct_ud_ep_last_skb_has_ack_req(ep) ||
+               /* Even if ACK_REQ message was already asked on last skb, post
+                * it in case of flush(CANCEL) to make sure sending it with
+                * SIGNALED flag set to get a corresponding CQE */
+               ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
```

Comment:
maybe inverse the order of if?
this way would need to check flush(cancel) only once:
```c
if (!uct_ud_ep_last_skb_has_ack_req(ep) || ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
  . ..
} else if (uct_ud_ep_is_last_ack_received(ep)) {
  /* no need to flush cancel - it's definitely false here, otherwise we would go to first if */
  ..
}
```

Reply:
unfortunately, we couldn't do it, since we should check `uct_ud_ep_last_skb_has_ack_req` only if `uct_ud_ep_is_last_ack_received` is true.
btw, I missed setting `UCT_UD_SEND_SKB_FLAG_ACK_REQ` - fixed it now.
also, I proposed another way on how we could have checks - for me it looks simpler